### ISSUE TITLE
Migrate diff report generation steps to script file

### ISSUE THREAD

**Issue by nrmancuso** [checkstyle/checkstyle#12991]
Noticed at https://github.com/checkstyle/checkstyle/pull/12990, we keep a lot of complex steps/logic in the https://github.com/checkstyle/checkstyle/blob/8a2aa986c75fd7ef136a51f87ff5ff751db49dc4/.github/workflows/diff-report.yml file. We should move steps out to a single script file, where we can invoke the script with a single argument command to handle steps, like we do with `validation.sh`.

This will help to clean up and make the yaml more readable, and allow us to run the same shellcheck version on the bash in this workflow that we do on other bash scripts, instead of relying on built-in shellcheck of actionlint.

Work on this migration has begin at https://github.com/checkstyle/checkstyle/pull/12990, we can continue to add steps as options in `diff-report.sh`.

**Comment by nrmancuso** [comment]
I am removing approved label and adding "discussion ongoing" label, we need to consider restructuring the entire workflow before we do this or instead of this issue. Report generation workflow should go like:
1. parse body of PR description, validate and download configs/projects files - this is the most frequent point of failure, so we should do this first to fail job as quickly as possible
2. checkout checkstyle repo
3. create temporary folder for all report generation, etc. called `ci-temp` to match other CI jobs
4. checkout contribution to that folder
5. generate reports
6. upload to s3
7. Place comment in PR

Main idea here is that we will wok from the root of the checksytle project, and not the workspace.

@romani @rnveach if you agree with this, I will create a new issue or just modify this one.

**Comment by romani** [comment]
Yes, all logical. I thought it worked like this already, but feel free to fix all you want.

**Comment by nrmancuso** [comment]
I am on it

**Comment by romani** [comment]
workflow https://github.com/checkstyle/checkstyle/blob/master/.github/workflows/regression-report.yml
started to use majority of proposed (download is not first).

For now such workflows are almost identical in functionality. both can leave in same time or one will survive.

**Comment by stoyanK7** [comment]
Curious, why create [`.github/workflows/regression-report.yml`](https://github.com/checkstyle/checkstyle/blob/master/.github/workflows/regression-report.yml) when  [`.github/workflows/diff-report.yml`](https://github.com/checkstyle/checkstyle/blob/master/.github/workflows/diff-report.yml) exists?

**Comment by romani** [comment]
.github/workflows/regression-report.yml focused on pre created configs at https://github.com/checkstyle/test-configs ,  https://github.com/checkstyle/test-configs/tree/main/AbbreviationAsWordInName/all-examples-in-one#all-examples-in-one-configs , so most cases, users don't need to make special config, we already have bunch pre-created. This workflow also covers big part of diff-report.yml, not all.

diff-report.yml is still good tool for new Checks, and very special execution, very configurable.

**Comment by DakshRJain737** [comment]
I am on it.

**Comment by avadhutmali** [comment]
Is this issue still open for contribution? Also, should we continue with the original plan of moving logic to a script, or align it with regression-report.yml?
